generator/src/swam/generator/ImportGenerator.scala

```diff
@@ -0,0 +1,150 @@
+    case I32 => s"Int"
+    case I64 => s"Long"
+    case F32 => s"Float"
+    case F64 => s"Double"
+  }
+
+  /**
+    * Change the template file
+    * @param file
+    */
+  def changeTemplate(file: String) = templateFile = file
+
+  /**
+    * Creates the string representing the scala function return based on the arity of the return object in WASM
+    * @param t
+    * @return
+    */
+  private def buildReturn(t: Vector[ValType]): String = {
+    if (t.isEmpty)
+      "Unit"
+    else {
+      if (t.length == 1)
+        getScalaType(t.last)
+      else
+        t.map(getScalaType).mkString("(", ",", ")")
+    }
+  }
+
+  /**
+    * Generates a context Map to feed the template rendering
+    * @param imports
+    * @return
+    */
+  def getContext(imports: Vector[Import]): Seq[Map[String, Any]] = {
+    val sorted = imports
+      .collect { case i: Import.Function => i } // Filter by function type
+      .groupBy(t => t.moduleName) // Group by module
+      .map[String, Set[Import.Function]](t => (t._1, t._2.toSet)) // remove duplicated entries
+
+    // Generating DTO
+    val dto = sorted.keys.zipWithIndex
+      .map(
+        m =>
+          Map(
+            "module" -> m._1,
+            "comma" -> (m._2 < sorted.keys.size - 1),
+            "fields" -> sorted(m._1).toSeq.zipWithIndex
+              .map(
+                f => {
+                  Map(
+                    "name" -> f._1.fieldName,
+                    "return" -> buildReturn(f._1.tpe.t),
+                    "params" -> f._1.tpe.params.zipWithIndex.map(t => s"p${t._2}:${getScalaType(t._1)}").mkString(","),
+                    "comma" -> (f._2 < sorted(m._1).size - 1)
+                  )
+                }
+              )
+          ))
+      .toSeq
+
+    dto
+  }
+
+  /**
+    * Creates the trait to implement the WASM import functions
+    * @param imports
+    * @return
+    */
+  def generateImportText(imports: Vector[Import]) = {
+
+    val te = new TemplateEngine()
+    te.boot()
+
+    val result =
+      te.layout(templateFile,
+                Map(
+                  "imports" -> getContext(imports)
+                ))
+
+    val file = Paths.get(s"GeneratedImport.scala")
+
+    try {
+      scalafmt.format(config, file, result)
+    } catch {
+      case x: Exception => {
+        throw new RuntimeException(s"It seems that the generated file is not a valid scala file. ${x.getMessage}")
+      }
+    }
+  }
+
+  /**
+    * Creates a new module including a trait object to implement the Imports
+    * @param projectName
+    * @param imports
+    * @return
+    */
+  def createScalaProjectForImports(projectName: String, imports: Vector[Import]) = {
+    s"mkdir -p $projectName".!!
+    s"mkdir -p $projectName/src".!!
+
+    val trait_ = generateImportText(imports)
+
+    val printer = new PrintWriter(new File(s"$projectName/src/GeneratedImport.scala"))
+    printer.print(trait_)
+    printer.close()
```

Comment:
Again, using fs2-io here to write in the file would be more reliable.

Reply:
Moved to fs2io